docs/src/user_docs/assembly/execution_contexts.md

```diff
@@ -0,0 +1,105 @@
+## Execution contexts
+Miden assembly program execution can span multiple isolated contexts. An execution context defines its own memory space which is not accessible from other execution contexts.
+
+All programs start executing in a *root* context. Thus, the main procedure of a program is always executed in the root context. To move execution into a different context, we can invoke a procedure using the `call` instruction. In fact, any time we invoke a procedure using the `call` instruction, the procedure is executed in a new context. We refer to all non-root contexts as *user contexts*.
+
+While executing in a user context, we can request some procedures to be executed in the root context. This can be done via the `syscall` instruction. The set of procedures which can be invoked via the `syscall` instruction is limited by a [kernel](#kernels) against which a program is compiled. Once the procedure called via `syscall` returns, the execution moves back to the user context from which it was invoked. The diagram below illustrates this graphically:
+
+![context transitions](../../assets/user_docs/assembly/execution_contexts/context_transitions.png)
+
+### Procedure invocation semantic
+As mentioned in the [previous section](./code_organization.md), procedures in Miden assembly can be invoked via 3 different instructions: `exec`, `call`, and `syscall`. Invocation semantics of `call` and `syscall` instructions are basically the same, the only difference being that the `syscall` instruction can be used only with procedures which are defined in the program's kernel. The `exec` instruction is different, and we explain these differences below.
+
+#### Invoking via `call` and `syscall` instructions
+When a procedure is invoked via a `call` or a `syscall` instruction, the following things happen:
+* Execution moves into a different context. In case of a `call` instruction, a new user context is created. In case of a `syscall` instruction, the execution moves back into the root context.
+* All stack items beyond the 16th item get "hidden" from the invoked procedure. That is, from the standpoint of the invoked procedure the stack depth after a `call` or a `syscall` is 16.
```

Comment:
I think the second sentence is a bit confusing, because we're taking the perspective of the invoked procedure but talking about after the `call`, at which point the procedure no longer has a perspective because it's done (unless I misunderstood).

Reply:
Reworded! (hopefully, for the better :) )